diff --git a/node_modules/optimism/lib/bundle.cjs.js b/node_modules/optimism/lib/bundle.cjs.js
index d5ab82f..5d68de1 100644
--- a/node_modules/optimism/lib/bundle.cjs.js
+++ b/node_modules/optimism/lib/bundle.cjs.js
@@ -227,7 +227,13 @@ var Entry = /** @class */ (function () {
     Entry.prototype.forgetDeps = function () {
         var _this = this;
         if (this.deps) {
-            toArray(this.deps).forEach(function (dep) { return dep.delete(_this); });
+            toArray(this.deps).forEach(function (dep) {
+                var result = dep.delete(_this);
+                if (dep["cleanup"]) {
+                    dep["cleanup"]();
+                }
+                return result;
+            });
             this.deps.clear();
             emptySetPool.push(this.deps);
             this.deps = null;
@@ -413,6 +419,13 @@ function dep(options) {
             var dep_1 = depsByKey.get(key);
             if (!dep_1) {
                 depsByKey.set(key, dep_1 = new Set);
+                if (global.enableOptimismFix) {
+                    dep_1.cleanup = () => {
+                        if (dep_1 && dep_1.size === 0 && dep_1 === depsByKey.get(key)) {
+                            depsByKey.delete(key);
+                        }
+                    }
+                }
             }
             parent.dependOn(dep_1);
             if (typeof subscribe === "function") {
diff --git a/node_modules/optimism/lib/bundle.esm.js b/node_modules/optimism/lib/bundle.esm.js
index 9e5a669..bfeacc4 100644
--- a/node_modules/optimism/lib/bundle.esm.js
+++ b/node_modules/optimism/lib/bundle.esm.js
@@ -225,7 +225,13 @@ var Entry = /** @class */ (function () {
     Entry.prototype.forgetDeps = function () {
         var _this = this;
         if (this.deps) {
-            toArray(this.deps).forEach(function (dep) { return dep.delete(_this); });
+            toArray(this.deps).forEach(function (dep) {
+                const result = dep.delete(_this);
+                if (dep["cleanup"]) {
+                    dep["cleanup"]();
+                }
+                return result;
+            });
             this.deps.clear();
             emptySetPool.push(this.deps);
             this.deps = null;
@@ -412,6 +418,14 @@ function dep(options) {
             if (!dep_1) {
                 depsByKey.set(key, dep_1 = new Set);
             }
+            if (self.enableOptimismFix) {
+                dep_1.cleanup = () => {
+                    if (dep_1 && dep_1.size === 0 && dep_1 === depsByKey.get(key)) {
+                        depsByKey.delete(key);
+                    }
+                }
+            }
+
             parent.dependOn(dep_1);
             if (typeof subscribe === "function") {
                 maybeUnsubscribe(dep_1);
@@ -432,6 +446,7 @@ function dep(options) {
             maybeUnsubscribe(dep);
         }
     };
+
     return depend;
 }
 
